File: examples/x2_oop_vehicle.py
#!/usr/bin/env python3
'''
@author  Michele Tomaiuolo - http://www.ce.unipr.it/people/tomamic
@license This software is free - http://www.gnu.org/licenses/gpl.html
'''

import g2d

ARENA_W, ARENA_H = 480, 360

class Vehicle:
    def __init__(self, pos: (int, int), dx: int):
        self._x, self._y = pos
        self._dx = dx
        self._left, self._right = -100, ARENA_W + 100

    def move(self):
        if self._x + self._dx < self._left:
            self._x = self._right
        if self._x + self._dx > self._right:
            self._x = self._left
        self._x += self._dx

    def uturn(self):
        self._dx *= -1

    def position(self):
        return self._x, self._y


def tick():
    if g2d.key_pressed("LeftButton"):
        b.uturn()
    g2d.clear_canvas()
    b.move()
    g2d.draw_image("ball.png", b.position())

def main():
    global b
    b = Vehicle((40, 40), 5)
    g2d.init_canvas((ARENA_W, ARENA_H))
    g2d.main_loop(tick)

main()

File: go/x2_oop_vehicle.go
package main

import . "g2d"

var screen = Point{480, 360}
var margin = 100
var a = NewVehicle(Point{40, 40}, 5)

type Vehicle struct {
    x, y, w, h      int
    dx, left, right int
}

func NewVehicle(pos Point, dx int) *Vehicle {
    return &Vehicle{pos.X, pos.Y, 20, 20, dx, -margin, screen.X+margin}
}

func (a *Vehicle) Move() {
    if a.x + a.dx < a.left {
        a.x = a.right
    }
    if a.x + a.dx > a.right {
        a.x = a.left
    }
    a.x += a.dx
}

func (a *Vehicle) Position() Point {
    return Point{a.x, a.y}
}

func (a *Vehicle) Uturn() {
    a.dx *= -1
}

func tick() {
    if KeyPressed("Enter") {
        a.Uturn()
    }
    ClearCanvas()
    a.Move()
    DrawImage(a.Position())
}

func main() {
    InitCanvas(screen)
    MainLoop(tick)
}

